src/web/middlewares/web_middleware_csrf.ml

```diff
@@ -1,6 +1,39 @@
-(* TODO
-https://docs.djangoproject.com/en/3.0/ref/csrf/#how-it-works *)
+let ( let* ) = Lwt.bind
 
-let m () =
-  let filter handler ctx = handler ctx in
-  Web_middleware_core.create ~name:"csrf" filter
+module CsrfService = struct
+  let add_to_ctx ctx ~token () =
+    let open Core_ctx in
+    let key : Token.t key = create_key () in
+    add key token ctx
+end
+
+module Make
+    (MessageService : Message.Sig.Service)
+    (TokenService : Token.Sig.SERVICE) =
+struct
+  let m () =
+    let filter handler ctx =
+      let open Lwt_result.Infix in
+      let* res =
+        Web_req.urlencoded ctx "csrf" >>= fun value ->
+        TokenService.find ctx ~value ()
+      in
+      let _ =
+        match res with
+        | Ok token -> TokenService.invalidate ctx ~token ()
+        | Error _ ->
+            Logs.err (fun m -> m "MIDDLEWARE: Failed to retrieve CSRF token");
+            MessageService.set ctx ~error:[ "TODO ERROR MESSAGE" ] ()
+      in
+      let* token = TokenService.create ctx ~kind:"csrf" () in
+      let ctx =
+        match token with
+        | Ok token -> CsrfService.add_to_ctx ctx ~token ()
+        | Error msg ->
```

Comment:
@jerben How to handle this error? Again a message?

Reply:
@aronerben If token creation fails, there is no way to recover from it. Throw an exception here, the web server shows error 500 to the user. We try to "throw and never catch" because you should only throw things that you can not recover from.

Reply:
Done